### DISCUSSION TITLE
How to load Uri in Jetpack Compose?

### DISCUSSION THREAD

**Discussion by sasaju** [coil-kt/coil#833] [Q&A]
- When I run this code, it can show picture correctly. ↓

```
Image(
    painter = rememberImagePainter(
        data = "android.resource://com.normalschedule.debug/drawable/main_background_4"
    ),
    contentDescription = null,
    modifier = Modifier.fillMaxSize(),
    contentScale = ContentScale.FillBounds
)
```

- But when I run this code, it only show white screen. ↓ (And I don't found errors in Logcat.)

```
Image(
    painter = rememberImagePainter(data = Uri.parse("content://com.android.externalstorage.documents/document/primary%3ADCIM%2FCamera%2F1577257945_5EC7B6D0109A60C8EC0CCA9413259D8C.jpg")),
    contentDescription = null,
    modifier = Modifier.fillMaxSize(),
    contentScale = ContentScale.FillBounds
)
```

- In order to make sure the path(_content://xxxx_) correctly, I use it in Glide. It can load correctly.

```
Image(
    painter = rememberGlidePainter(request = Uri.parse("content://com.android.externalstorage.documents/document/primary%3ADCIM%2FCamera%2F1577257945_5EC7B6D0109A60C8EC0CCA9413259D8C.jpg")),
    contentDescription = null,
    modifier = Modifier.fillMaxSize(),
    contentScale = ContentScale.FillBounds
)
```

- And `rememberCoilPainter()` has same problem.

**So, does coil support to load "content://xxx"? If support it what should I do?**

**Comment by sasaju** [comment]
I solve this problem by requesting READ_EXTERNAL_STORAGE permission. But I don't know why Glide can load it when don't have this permission.😑

**Reply by 809825706** [reply]
how to show a image with local file？I can show image from http://XXX but not content://com.android.externalstorage.documents/XXX

```kotlin
    val ht = "https://avatars2.githubusercontent.com/u/3265208?v=3&s=100"
    val uri = "content://com.android.externalstorage.documents/document/primary%3ADocuments%2Forgzly%2Flogs%2Ftemp%2FH3K27AC_diff-peaks.png"

        Image(
            painter = rememberImagePainter(
                data  = Uri.parse(uri)  // or ht
            )
            ,
            contentDescription = "123",
            modifier = Modifier.fillMaxSize(),
            contentScale = ContentScale.FillWidth
        )
```

when use uri, nothing was show.I have set:

```
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
```

```
    implementation "io.coil-kt:coil-compose:1.4.0"
```

**Reply by sasaju** [reply]
Are you give your app READ_EXTERNAL_STORAGE permissions?

**Reply by 809825706** [reply]
yes

![image](https://user-images.githubusercontent.com/46601791/138599106-ac5fb0dc-70a7-445a-a04d-e7b815a1b203.png)

**Reply by sasaju** [reply]
Emm....Maybe we face different situations. So, I can't give you the suitable suggestion. Sorry

**Reply by TippuFisal** [reply]
> how to show a image with local file？I can show image from http://XXX but not content://com.android.externalstorage.documents/XXX
> 
> ```kotlin
>     val ht = "https://avatars2.githubusercontent.com/u/3265208?v=3&s=100"
>     val uri = "content://com.android.externalstorage.documents/document/primary%3ADocuments%2Forgzly%2Flogs%2Ftemp%2FH3K27AC_diff-peaks.png"
> 
>         Image(
>             painter = rememberImagePainter(
>                 data  = Uri.parse(uri)  // or ht
>             )
>             ,
>             contentDescription = "123",
>             modifier = Modifier.fillMaxSize(),
>             contentScale = ContentScale.FillWidth
>         )
> ```
> 
> when use uri, nothing was show.I have set:
> 
> ```
>     <uses-permission android:name="android.permission.INTERNET" />
>     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
> ```
> 
> ```
>     implementation "io.coil-kt:coil-compose:1.4.0"
> ```

It works but rememberImagePainter is deprecated so use this [rememberAsyncImagePainter]

**Comment by CatalinStefan** [comment]
I had the same problem, and for me the issue was that Android resources contain an extra ":", or rather "%3A" character towards the end of the URI. It can be read correctly by Glide, but not by Coil.
So if you encode and decode your URI at some point, make sure to convert the final ":" character to "%3A", otherwise it fails.

Took me 2 days to find this out. It had nothing to do with permissions

**Comment by gaohomway** [comment]
Requires conversion and permissions, just need to get the Uri in the right way to display.
```
val contentUri = ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, cursor.getLong(id))
```

**Reply by carterhudson** [reply]
My issue was that the contact URI was not appended with `ContactsContract.Contacts.Photo.DISPLAY_PHOTO`, which Coil explicitly checks for, but Glide apparently does not.

**Reply by Waldmann375** [reply]
what is cursor here? where to take it from?

**Comment by sDevPrem** [comment]
I solved this issue by taking Persistable URI Permission when I get the URI after the user has picked the picture and before saving it into the database:
```
 context.contentResolver.takePersistableUriPermission(photo_uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)

```

**Comment by anonym24** [comment]
I'm using SAF so I don't need to use any legacy READ/WRITE permissions...

Please fix it..
